nabs/streams.py

```diff
@@ -0,0 +1,73 @@
+"""Standard data streams for automated routines"""
+import numpy as np
+from bluesky.callbacks.stream import LiveDispatcher
+
+
+class AverageStream(LiveDispatcher):
+    """
+    Stream that averages data points together
+
+    As event documents are emitted from the RunEngine they are collected by
+    AverageStream, averaged and re-emitted as a secondary event stream. This
+    allows callbacks to subscribe to the average stream for more elegant
+    visualization and analysis.
+
+    The average stream can either be constructed with a specific number of
+    points to average or this information can be passed through in the start
+    document metadata. See the :meth:`.start` for more information.
+
+    Parameters
+    ----------
+    n : int, optional
+        Number of points to average together
+    """
+    def __init__(self, n=None):
+        self.n = n
+        self.last_event = None
+        self.raw_cache = list()
+        super().__init__()
+
+    def start(self, doc):
+        """
+        Create the stream after seeing the start document
+
+        The callback looks for the 'average' key in the start document to
+        configure itself.
+        """
+        # Grab the average key
+        self.n = doc.get('average', self.n)
+        super().start(doc)
+
+    def event(self, doc):
+        """Send an Event through the stream"""
+        # Add event to raw cache
+        self.raw_cache.append(doc)
+        # If we have enough events average
+        if len(self.raw_cache) == self.n:
+            average_evt = dict()
+            # Check that all of our events came from the same configuration
+            desc_id = self.raw_cache[0]['descriptor']
+            if not all([desc_id == evt['descriptor']
+                        for evt in self.raw_cache]):
+                raise Exception('The events in this bundle are from '
+                                'different configurations!')
+            # Use the last descriptor to avoid strings and objects
+            data_keys = self.raw_descriptors[desc_id]['data_keys']
+            for key, info in data_keys.items():
+                # Information from non-number fields is dropped
+                if info['dtype'] in ('number', 'array'):
+                    # Average together
+                    average_evt[key] = np.mean([evt['data'][key]
+                                                for evt in self.raw_cache],
+                                               axis=0)
+            self.last_event = {'data': average_evt, 'descriptor': desc_id}
```

Comment:
Nice catch. `process_event` sets a new `timestamp` for the event that just corresponds to the time the calculation was done, as well as pointing to the new descriptor e.t.c. So to answer your question, the `last_event` is malformed until it is put through `process_event`.

Do you think we should return the whole event here? The old `measure_average` only cared about what was contained within "data"  so I didn't care about the other stuff as much. I figured if you wanted full events you would just `subscribe`. Maybe this was an oversight.

Reply:
So just store the `"data"` as `last_average`?

Reply:
Kind of leaning towards this. Clear delineation -> want context like descriptors and timestamps, multiple events you must subscribe. Want purely processed data for adaptive scans, grab it from the callback itself.

Reply:
Yeah, I like that, since the descriptor id doesn't mean anything anyway unless you're subscribing and get the descriptor doc, so as is it just adds an extra key to the workflow.